core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -198,15 +198,38 @@ private[spark] class ExecutorAllocationManager(
     t.start()
   }
 
+  def currentTargetNumExecutors(): Int =
+    numExecutorsPending + executorIds.size - executorsPendingToRemove.size
+
   /**
-   * If the add time has expired, request new executors and refresh the add time.
-   * If the remove time for an existing executor has expired, kill the executor.
+   * This is called at a fixed interval to relegate the number of pending executor requests
+   * and number of executors running.
+   *
+   * First, check to see whether our existing allocation and the requests we've made previously
+   * exceed our current needs. If so, let the cluster manager know so that it can cancel pending
+   * requests that are unneeded.
+   *
+   * If not, and the add time has expired, see if we can request new executors and refresh the add
+   * time.
+   *
+   * Last, if the remove time for an existing executor has expired, kill the executor.
+   *
    * This is factored out into its own method for testing.
    */
   private def schedule(): Unit = synchronized {
+    // The maximum number of executors we need under the current load is the total number of
+    // running or pending tasks, divided by the full task capacity of each executor, rounded up.
+    val numRunningOrPendingTasks = listener.totalPendingTasks + listener.totalRunningTasks
+    val maxNumNeededExecutors = (numRunningOrPendingTasks + tasksPerExecutor - 1) / tasksPerExecutor
+
     val now = clock.getTimeMillis
-    if (addTime != NOT_SET && now >= addTime) {
-      addExecutors()
+    val currentTarget = currentTargetNumExecutors
+    if (maxNumNeededExecutors < currentTarget) {
+      client.requestTotalExecutors(maxNumNeededExecutors)
+      numExecutorsPending -= math.min(currentTarget - maxNumNeededExecutors, 0)
```

Comment:
this looks incorrect. If `currentTarget > maxNumNeededExecutors` the min will always return 0.